app/models/capability.rb

```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+class Capability
+  include CanCan::Ability
+
+  def initialize(user, token)
+    if user
+      can [:read, :read_one], UserPreference if capability?(token, :allow_read_prefs)
+      can [:update, :update_one, :delete_one], UserPreference if capability?(token, :allow_write_prefs)
+
+    end
+  end
+
+  private
+
+  # If a user provides no tokens, they've authenticated via a non-oauth method
+  # and permission to access to all capabilities is assumed.
+  def capability?(token, cap)
+    token.nil? || token.read_attribute(cap)
```

Comment:
The original `token.nil?` check was here when you were combining both the Ability and Capability stuff, so for normal web stuff it was ok if the token was nil. But it's been refactored since, so that it's only used for token-based capability checks. I worry that this opens up a foot-gun situation, since as far as I can tell there's no reason for calling `capability?` without a token -  what do you think?